cmake/modules/kconfig.cmake

```diff
@@ -379,7 +379,21 @@ import_kconfig(${KCONFIG_NAMESPACE} ${DOTCONFIG})
 # Cache the CLI Kconfig symbols that survived through Kconfig, prefixed with CLI_.
 # Remove those who might have changed compared to earlier runs, if they no longer appears.
 foreach (name ${cli_config_list})
+  # For bool and tristate symbols, Kconfig accepts any value starting with
+  # 'y', 'm', 'n', but only the first character is used.
+  string(REGEX MATCH "^[ymn]|.*" value ${CLI_${name}})
   if(DEFINED ${name})
+    # For string symbols, import_kconfig() strips out quotation marks,
+    # but 'value' should still have them intact.
+    set(possible_values "${${name}}" "\"${${name}}\"")
+  else()
+    # For bool/tristate, import_kconfig() leaves (=n) symbols unset,
+    # but we want to preserve explicit assignments to "n" and avoid
+    # reverting to default values on a re-run.
+    set(possible_values "n")
+  endif()
+
+  if(value IN_LIST possible_values)
```

Comment:
I submitted an alternative solution inspired by your suggestion.

I agree that `import_kconfig()` could be made more robust, so I added another patch for that here. Thanks for the tips.

There's only one detail that I'm unsure about. In the case of either:
```
# CONFIG_FOO is not set
CONFIG_FOO=n
```
did you want the CMake variable `CONFIG_FOO` to be set to a literal `n`, rather than unset? Because that would be a breaking change.

Reply:
> I submitted an alternative solution inspired by your suggestion.

Great.

> There's only one detail that I'm unsure about. In the case of either:
> ```
> # CONFIG_FOO is not set
> CONFIG_FOO=n
> ```

The `CONFIG_FOO=n` version is not valid in the `.config` file.
If it is present, then kconfiglib will transform that line into `# CONFIG_FOO is not set`.

This means that the form `CONFIG_FOO=n` can only be present in `.config` if a user manually has modified the file, and in such case will be short lived as the next run of kconfiglib (which is triggered by CMake) will update the setting and hence the line will be `# CONFIG_FOO is not set` when `import_kconfig()` executes.

Also note, if a user manually modifies the `.config` then it's in principle the user's responsibility that the format is correct.

> did you want the CMake variable CONFIG_FOO to be set to a literal n, rather than unset? Because that would be a breaking change.

That is a valid point.
For consistency we should probably set it to `n` explicitly, and the way we generally use `CONFIG_<X>` in CMake it should work.
But as there could be occurrences where someone does a `if(DEFINED CONFIG_<X>)`, then keeping current behavior to unset is the safest approach, so let's stick to that with this PR.

I believe changing behavior is better done in a separate PR, where pros / cons can be discussed.

Reply:
Whilst that change would be nice, it's a bit "too little too late". If that change was made 5 years ago then it would have been fine, but to make that change _now_ would cause **major** breakage both in and out of tree, needing a lot of things to be reworked and I'm not in favour of that.

Reply:
> but to make that change now would cause major breakage both in and out of tree

Not sure I would consider it a major breakage.

If someone does `if(DEFINED CONFIG_FOO)` when they want to test if `FOO=y` or `if(NOT DEFINED CONFIG_FOO)` when the assume `FOO=n` they are doing it wrong.

If `FOO` is considered a boolean, then it ought to be checked as a bool, and not rely on a defined variable being equal to a true value.

But I acknowledge that to many developers then the difference between `if(CONFIG_FOO)` and `if(DEFINED CONFIG_FOO)` may go unnoticed as long as the CMake code behaves as they intend.